File: Python/Rank Up/Evaluate mathematical expression.py
# def isMathSign(symbol: str) -> bool:
#     """ Returns True if 'symbol' is a math sign. Otherwise - returns False """
#     if symbol == '*' or symbol == '/' or symbol == '+' or symbol == '-':
#         return True
#     return False


# def calcSimplestExpr(a: float, b: float, sign: str) -> float:
#     """ Returns calculated expression in relation to specified math sign. If 'sign' is unknown - returns None """
#     if sign == '*':
#         return a * b
#     elif sign == '/':
#         return a / b
#     elif sign == '+':
#         return a + b
#     elif sign == '-':
#         return a - b
#     else:
#         return None


# def replaceAllDoubleSigns(expr: str) -> str:
#     """ Replaces all double signs in the math expression """
#     return expr.replace(
#         '+-', '-').replace('-+', '-').replace('++', '+').replace('--', '+')


# def calcGenericPart(expr: str, posOfSign: int) -> str:
#     """ Performs calculation of simplest expression with modifying passed string,
#     i.e. previous expression erasing and replaced with result of the calculation """

#     sign = expr[posOfSign]
#     isSndNumIsNeg = False

#     posOfLastDigInFstNum = posOfSign - 1
#     posOfFirstDigInFstNum = posOfLastDigInFstNum
#     posOfFirstDigInSndNum = posOfSign + 1
#     posOfLastDigInSndNum = posOfFirstDigInSndNum

#     # Defining first digit of the first number
#     while not isMathSign(expr[posOfFirstDigInFstNum]):
#         # Last iteration -> sign
#         if posOfFirstDigInFstNum != 0:
#             posOfFirstDigInFstNum -= 1

#         if (expr[posOfFirstDigInFstNum] == 'e'):
#             posOfFirstDigInFstNum += 1

#         if posOfFirstDigInFstNum == 0 or posOfFirstDigInFstNum < 0:
#             break

#     if posOfFirstDigInFstNum != 0:
#         posOfFirstDigInFstNum += 1

#     if isMathSign(expr[posOfFirstDigInSndNum]):
#         posOfFirstDigInSndNum += 1
#         posOfLastDigInSndNum += 1
#         isSndNumIsNeg = True

#     # Defining last digit in the second number
#     while not isMathSign(expr[posOfLastDigInSndNum]):
#         # Last iteration -> sign
#         posOfLastDigInSndNum += 1

#         if posOfLastDigInSndNum == len(expr):
#             break

#         if (expr[posOfLastDigInSndNum] == 'e'):
#             posOfLastDigInSndNum += 2

#     if posOfLastDigInSndNum != len(expr):
#         posOfLastDigInSndNum -= 1

#     fstNum = float(expr[posOfFirstDigInFstNum:posOfLastDigInFstNum + 1])
#     sndNum = float(expr[posOfFirstDigInSndNum:posOfLastDigInSndNum + 1])

#     if isSndNumIsNeg:
#         sndNum *= -1

#     # Replacing old numbers from the string and inserting new calculated number
#     expr = expr.replace(expr[posOfFirstDigInFstNum:posOfLastDigInSndNum + 1],
#                         str(calcSimplestExpr(fstNum, sndNum, sign)))

#     # Getting rid of unnecessary repetitive characters
#     expr = replaceAllDoubleSigns(expr)

#     return expr


# def calcSpecificPart(expr: str, countOfSigns: int, signs: bool = True) -> str:
#     """ Returns calculated part of an expression in dependence of specified signs
#         If bool flag 'signs' is "true" -> operations are "*" and "/"
#         Otherwise - "+" and "-" """

#     sign1 = str()
#     sign2 = str()

#     # At first we calculate in order what is multiplied or divided, then the rest
#     if signs == True:
#         sign1 = '*'
#         sign2 = '/'
#     else:
#         sign1 = '+'
#         sign2 = '-'

#     i = int()
#     while (i < countOfSigns):
#         posOfSign = int()
#         for j, ch in enumerate(expr):
#             if expr[j] == sign1 or expr[j] == sign2:
#                 posOfSign = j

#                 if posOfSign != 0:
#                     break
#         expr = calcGenericPart(expr, posOfSign)
#         i += 1

#     return expr


# def calcWholeExpression(expr: str) -> str:
#     """ Returns calculated expression from the one expression """
#     # Initializing copy of the initial expression to erasing unnecessary symbols
#     exprCopy = expr

#     # Counting multiplication and divion sings
#     countOfMultDiv = sum(1 for x in expr if x == '*' or x == '/')

#     # We need to calculate the expression count of signs times
#     # so, we have count of the multiplication and division signs
#     # which we would calculate at first in the order which they
#     # are located in the expression, then - calculate the rest of
#     # add and subtracting operations
#     exprCopy = calcSpecificPart(exprCopy, countOfMultDiv)

#     # Getting rid of unnecessary repetitive characters
#     expr = replaceAllDoubleSigns(expr)

#     # Counting adding and subtracting sings
#     countOfAddSubtr = sum(1 for x in exprCopy if x == '+' or x == '-')

#     if exprCopy.find('+') == 0 or exprCopy.find('-') == 0:
#         countOfAddSubtr -= 1

#     if exprCopy.find('E+') != -1 or exprCopy.find('e+') != -1 or exprCopy.find('E-') != -1 or exprCopy.find('e-') != -1:
#         countOfAddSubtr -= 1

#     exprCopy = calcSpecificPart(exprCopy, countOfAddSubtr, False)

#     return exprCopy


# def calc(expr: str) -> float:
#     """ Calculates expression from the string representation """

#     # If we got an empty string - return nil as a result
#     if expr == '' or expr == ' ':
#         return 0.0

#     # At first, removing all whitespaces from the string and replacing all double signs
#     expr = expr.replace(' ', '')
#     expr = replaceAllDoubleSigns(expr)

#     # Finding position of brackets
#     posOflbracket = expr.rfind('(')
#     posOfrbracket = expr.find(')')

#     # If there is no more brackets
#     while posOflbracket != -1 and posOfrbracket != -1:
#         # Updating positions for the next calculations
#         posOflbracket = expr.rfind('(')
#         posOfrbracket = expr.find(')')

#         if posOflbracket == -1 and posOfrbracket == -1:
#             break

#         exprInParentheses = expr[posOflbracket + 1:posOfrbracket]

#         # Erasing expression in parentheses
#         expr = expr.replace('(' + exprInParentheses + ')', '')
#         # Replacing erased expression in parentheses with a calculation result
#         expr = expr[:posOflbracket] + \
#             calcWholeExpression(exprInParentheses) + expr[posOflbracket:]
#         # Getting rid of unnecessary repetitive characters
#         expr = replaceAllDoubleSigns(expr)

#     # If expression doesn't have a parentheses - just calculate whole expression
#     if posOflbracket == -1 and posOfrbracket == -1:
#         calculated = calcWholeExpression(expr)

#         # Replacing expression in parentheses with a calculation result
#         expr = calculated

#     return float(expr)


# print(calc('1 + 1'))
# print(calc('8/16'))
# print(calc('3 -(-1)'))
# print(calc('2 + -2'))
# print(calc('10- 2- -5'))
# print(calc('(((10)))'))
# print(calc('3 * 5'))
# print(calc('-7 * -(6 / 3)'))
# print(calc('61 / -9 / -17 * -3 - 40 * -73 - 23 * -12'))
# print(calc('-34 + 45 - 5 * -73 * -28 - 79 - -28 - 31'))
# print(calc('-71 - -76 - -19 + -84 / 42 / -96 / -66 / 99'))
# print(calc('19-3.1884501581471274e-06'))


### BELOW IS PRESENTED NOT MY SOLUTION !!! ###
### MY SOLUTION IS ABOVE - ALL COMMENTED CODE ###
### BUT IT CALLS TIME LIMIT (more than 12s) ###
import re
from operator import mul, truediv as div, add, sub


OPS = {'*': mul, '/': div, '-': sub, '+': add}


def calc(expression):
    tokens = re.findall(r'[.\d]+|[()+*/-]', expression)
    return parse_AddSub(tokens, 0)[0]


def parse_AddSub(tokens, iTok):

    v, iTok = parse_MulDiv(tokens, iTok)

    while iTok < len(tokens) and tokens[iTok] != ')':
        tok = tokens[iTok]
        if tok in '-+':
            v2, iTok = parse_MulDiv(tokens, iTok+1)
            v = OPS[tok](v, v2)

    return v, iTok


def parse_MulDiv(tokens, iTok):

    v, iTok = parse_Term(tokens, iTok)

    while iTok < len(tokens) and tokens[iTok] in '*/':
        op = tokens[iTok]
        v2, iTok = parse_Term(tokens, iTok+1)
        v = OPS[op](v, v2)

    return v, iTok


def parse_Term(tokens, iTok):
    tok = tokens[iTok]

    if tok == '(':
        v, iTok = parse_AddSub(tokens, iTok+1)
        if iTok < len(tokens) and tokens[iTok] != ')':
            raise Exception()

    elif tok == '-':
        v, iTok = parse_Term(tokens, iTok+1)
        v, iTok = -v, iTok-1

    else:
        v = float(tok)

    return v, iTok+1

File: C++/Rank Up/Evaluate mathematical expression.cpp
// Task Instructions:
// Given a mathematical expression as a string you must return the result as a number.

// Numbers:
// Number may be both whole numbers and/or decimal numbers. The same goes for the returned result.
// Operators

// You need to support the following mathematical operators:
// Multiplication *
// Division / (as floating point division)
// Addition +
// Subtraction -

// Operators are always evaluated from left-to-right, and * and / must be evaluated before + and -.

// Parentheses:
// You need to support multiple levels of nested parentheses, ex. (2 / (2 + 3.33) * 4) - -6

// Whitespace:
// There may or may not be whitespace between numbers and operators.
// An addition to this rule is that the minus sign (-) used for negating numbers
// and parentheses will never be separated by whitespace. I.e all of the following are valid expressions.

// 1-1    // 0
// 1 -1   // 0
// 1- 1   // 0
// 1 - 1  // 0
// 1- -1  // 2
// 1 - -1 // 2
// 1--1   // 2

// 6 + -(4)   // 2
// 6 + -( -4) // 10

// And the following are invalid expressions:
// 1 - - 1    // Invalid
// 1- - 1     // Invalid
// 6 + - (4)  // Invalid
// 6 + -(- 4) // Invalid

// Validation:
// You do not need to worry about validation - you will only receive
// valid mathematical expressions following the above rules.

#include <iostream>
#include <string>
#include <algorithm>
#include <iterator>

// Simple lambda expr that defines if 'ch' symbol is mathematical operation sign
constexpr auto isMathSign{[](char const &ch)
                          { return ch == '*' || ch == '/' || ch == '+' || ch == '-'; }};

// Returns calculated number from two numbers and sign represented as "std::string"
std::string calcNum(double const &a, double const &b, char const &op)
{
    switch (op)
    {
    case '*':
        return std::to_string(a * b);
    case '/':
        return std::to_string(a / b);
    case '+':
        return std::to_string(a + b);
    case '-':
        return std::to_string(a - b);
    default:
        return "";
    }
}

// Returns calculated part of an expression
std::string calcGenericPart(std::string &expr, size_t const &posOfSign)
{
    char sign{expr.at(posOfSign)};
    bool isSndNumberNegative{false};
    size_t posOfLastDigInFstNum{posOfSign - 1UL}, posOfFirstDigInFstNum{posOfLastDigInFstNum},
        posOfFirstDigInSndNum{posOfSign + 1UL}, posOfLastDigInSndNum{posOfFirstDigInSndNum};

    // Defining first digit of the first number
    while (!isMathSign(expr.at(posOfFirstDigInFstNum)))
    {
        // Last iteration -> sign
        if (posOfFirstDigInFstNum != 0UL)
            posOfFirstDigInFstNum--;

        if (posOfFirstDigInFstNum == 0UL)
            break;
    }

    if (posOfFirstDigInFstNum != 0UL)
        posOfFirstDigInFstNum++;

    if (isMathSign(expr.at(posOfFirstDigInSndNum)))
    {
        posOfFirstDigInSndNum++;
        posOfLastDigInSndNum++;
        isSndNumberNegative = true;
    }

    // Defining last digit of the second number
    while (!isMathSign(expr.at(posOfLastDigInSndNum)))
    {
        // Last iteration -> sign
        posOfLastDigInSndNum++;
        if (posOfLastDigInSndNum == expr.length())
            break;
    }

    if (posOfLastDigInSndNum != expr.length())
        posOfLastDigInSndNum--;

    // Converting strings to the double numbers
    double fstNum{std::stod(expr.substr(posOfFirstDigInFstNum, posOfSign - posOfFirstDigInFstNum))},
        sndNum{std::stod(expr.substr(posOfFirstDigInSndNum,
                                     posOfLastDigInSndNum - posOfFirstDigInSndNum + 1UL))};

    if (isSndNumberNegative)
        sndNum *= -1.0;

    // Erasing old numbers and inserting new calculated number
    expr.erase(posOfFirstDigInFstNum, posOfLastDigInSndNum - posOfFirstDigInFstNum + 1UL);
    expr.insert(posOfFirstDigInFstNum, calcNum(fstNum, sndNum, sign));

    return expr;
}

// Returns calculated part of an expression in dependence of specified signs
// If bool flag 'signs' is "true" -> operations are "*" and "/"
// Otherwise - "+" and "-"
std::string calcSpecificPart(std::string &expr, size_t const &countOfSigns,
                             bool const &signs = true)
{
    char sign1{}, sign2{};

    // At first we calculate in order what is multiplied or divided, then the rest
    if (signs == true)
        sign1 = '*', sign2 = '/';
    else
        sign1 = '+', sign2 = '-';

    for (size_t i{}; i < countOfSigns; i++)
    {
        size_t posOfSign{};
        for (size_t j{}; j < expr.size(); j++)
        {
            if (expr.at(j) == sign1 || expr.at(j) == sign2)
            {
                posOfSign = j;

                if (posOfSign != 0UL)
                    break;
            }

            if (expr.find(sign1) == std::string::npos &&
                expr.find(sign2) == std::string::npos)
                break;
        }
        calcGenericPart(expr, posOfSign);
    }
    return expr;
}

// Modifies expression if expression has double signs
void replaceDoubleSignsInExpr(std::string &expr, std::string const &signs)
{
    size_t posOfDoubleSigns{expr.find(signs)};

    // While ".find()" do not return "std::string::npos" (max size of the string)
    while (posOfDoubleSigns != std::string::npos)
    {
        if (signs == "--" || signs == "++")
            expr.replace(expr.find(signs), 2UL, "+");
        else if (signs == "+-" || signs == "-+")
            expr.replace(expr.find(signs), 2UL, "-");

        // Updating position of signs
        posOfDoubleSigns = expr.find(signs);
    }
}

// Returns calculated expression from the one expression
std::string calcWholeExpression(std::string const &expr)
{
    // Initializing copy of the initial expression to erasing unnecessary symbols
    std::string exprCopy{expr};

    // Counting multiplication and divion sings
    auto countOfMultDiv{std::count_if(std::cbegin(exprCopy), std::cend(exprCopy),
                                      [](char const &sign)
                                      {
                                          return sign == '*' || sign == '/';
                                      })};

    // We need to calculate the expression count of signs times
    // so, we have count of the multiplication and division signs
    // which we would calculate at first in the order which they
    // are located in the expression, then - calculate the rest of
    // add and subtracting operations
    calcSpecificPart(exprCopy, static_cast<size_t>(countOfMultDiv));

    // Getting rid of unnecessary repetitive characters
    replaceDoubleSignsInExpr(exprCopy, "--");
    replaceDoubleSignsInExpr(exprCopy, "-+");
    replaceDoubleSignsInExpr(exprCopy, "+-");
    replaceDoubleSignsInExpr(exprCopy, "++");

    // Counting adding and subtracting sings
    auto countAddSubtr{std::count_if(std::cbegin(exprCopy), std::cend(exprCopy),
                                     [](char const &sign)
                                     {
                                         return sign == '+' || sign == '-';
                                     })};

    // Sign of negative number doesn't count
    if (exprCopy.find('+') == 0UL || exprCopy.find('-') == 0UL)
        countAddSubtr--;

    calcSpecificPart(exprCopy, static_cast<size_t>(countAddSubtr), false);

    return exprCopy;
}

double calc(std::string expr)
{
    // If we got an empty string - return nil as a result
    if (expr.empty() || expr == " ")
        return 0.0;

    // Removing all whitespaces from the expression
    expr.erase(std::remove_if(std::begin(expr), std::end(expr), [](char const &ch)
                              { return ch == ' '; }),
               std::cend(expr));

    // Finding position of brackets
    size_t posOflbracket{expr.find_last_of('(')}, posOfrbracket{expr.find_first_of(')')};

    // If there is no more brackets
    while (posOflbracket != std::string::npos && posOfrbracket != std::string::npos)
    {
        // Updating positions for the next calculations
        posOflbracket = expr.find_last_of('(');
        posOfrbracket = expr.find_first_of(')');

        if (posOflbracket == std::string::npos && posOfrbracket == std::string::npos)
            break;

        std::string exprInParentheses{expr.substr(posOflbracket + 1UL,
                                                  posOfrbracket - posOflbracket - 1UL)};

        // Erasing expression in parentheses
        expr.erase(posOflbracket, exprInParentheses.length() + 2UL);

        // Replacing erased expression in parentheses with a calculation result
        expr.insert(posOflbracket, calcWholeExpression(exprInParentheses));
    }

    // If expression doesn't have a parentheses - just calculate whole expression
    if (posOflbracket == std::string::npos && posOfrbracket == std::string::npos)
    {
        std::string calculated{calcWholeExpression(expr)};

        // Erasing all symbols in the expression
        expr.clear();

        // Replacing erased expression in parentheses with a calculation result
        expr.insert(0UL, calculated);
    }
    return std::stold(expr);
}

int main()
{
    std::cout << calc("150.000000-450") << std::endl;
    std::cout << calc("150.000000450") << std::endl;
    std::cout << calc("-1.335000-150.000000-450") << std::endl;
    std::cout << calc("2+1.665000-5-600.000000/4-450") << std::endl;
    std::cout << calc("(2 / (2 + 3.33 / 4 * 2 - 5 - 6 * 100 / 4 - 450) * 4) - -6") << std::endl;
    std::cout << calc("1 + 1") << std::endl;
    std::cout << calc("2 + 3.33 / 4 * 2 - 5 - 6 * 100 / 4 - 450") << std::endl;
    std::cout << calc("100") << std::endl;
    std::cout << calc("2 + 2 * 2") << std::endl;
    std::cout << calc("2 + 2 / 2 * 2") << std::endl;
    std::cout << calc("(2 / (2 + 3.33) * 4) - -6") << std::endl;
    std::cout << calc("(           (2/-   601.335000)*4)+6") << std::endl;
    std::cout << calc("-0.013304+6") << std::endl;
    std::cout << calc("8/16") << std::endl;
    std::cout << calc("3 -(-1)") << std::endl;
    std::cout << calc("2 + -2") << std::endl;
    std::cout << calc("10- 2- -5") << std::endl;
    std::cout << calc("(((10)))") << std::endl;
    std::cout << calc("3 * 5") << std::endl;
    std::cout << calc("-7 * -(6 / 3)") << std::endl;
    std::cout << calc("-1") << std::endl;
    std::cout << calc("-(-1)") << std::endl;
    std::cout << calc("-(-(1))") << std::endl;
    std::cout << calc("10 * 24 * -24 * 36 * -2 + -37 / -89 * 83") << std::endl;
    std::cout << calc("-72 - -21 / 94 - -85 / -2 / -31 * -64 - -30") << std::endl;
    std::cout << calc("") << std::endl;
    std::cout << calc(" ") << std::endl;
    std::cout << calc("0") << std::endl;
    std::cout << calc(" 0 ") << std::endl;
    std::cout << calc("-16 / -42 / -42 - -94 - 12 + 76 / -89 * -100") << std::endl;
    std::cout << calc("77 / 83 / 82 * -7 * -10 - 38 - -35 - 37") << std::endl;
    std::cout << calc("-(-(-1))") << std::endl;

    return EXIT_SUCCESS;
}

// Tests:
/*
Expected: equal to 195.394 (+/- 1e-06) -> Actual: 195.394
Expected: equal to -225224 (+/- 1e-06) -> Actual: -225224
Expected: equal to 410.416 (+/- 1e-06) -> Actual: 410.416
Expected: equal to 307.194 (+/- 1e-06) -> Actual: 307.194
Expected: equal to -6682.63 (+/- 1e-06) -> Actual: -6682.63
Expected: equal to 222.811 (+/- 1e-06) -> Actual: 222.811
Expected: equal to -128.996 (+/- 1e-06) -> Actual: -128.996
Expected: equal to 56.8048 (+/- 1e-06) -> Actual: 56.8048
Expected: equal to 109.772 (+/- 1e-06) -> Actual: 109.771
Expected: equal to -5762.62 (+/- 1e-06) -> Actual: -5762.62
*/
